src/InlineAutocomplete/utils.ts

```diff
@@ -0,0 +1,156 @@
+const multiWordTriggerTerminators = new Set(['.', '\n'])
+
+const isWhitespace = (char: string) => /\s/.test(char)
+
+/**
+ * Calculate whether or not suggestions should be shown based on the given state of the
+ * input. If they should be shown, returns the show event.
+ */
+export const calculateSuggestionsQuery = (
+  triggers: Array<Trigger>,
+  text: string,
+  caretLocation: number
+): ShowSuggestionsEvent | null => {
+  // Build backwards from the caret location until the most recent trigger character or terminator
+  for (
+    let i = caretLocation - 1, query = '', potentialTriggers = triggers;
+    i >= 0 && potentialTriggers.length > 0;
+    i--
+  ) {
+    const character = text[i]
+
+    if (singleWordTriggerTerminators.has(character)) potentialTriggers = potentialTriggers.filter(t => t.multiWord)
+    if (multiWordTriggerTerminators.has(character)) potentialTriggers = potentialTriggers.filter(t => !t.multiWord)
+
+    for (const trigger of potentialTriggers.filter(t => character === t.triggerChar)) {
+      // Trigger chars must always be preceded by whitespace or be the first character in the input,
+      // and even a multi-word query cannot start with whitespace
+      if ((i === 0 || isWhitespace(text[i - 1])) && !isWhitespace(query[0])) return {trigger, query}
+
+      potentialTriggers = potentialTriggers.filter(t => t !== trigger)
+    }
+
+    query = character + query
+  }
+
+  return null
+}
+
+/**
+ * Obtain the coordinates (px) of the bottom left of a character in an input, relative to the
+ * top-left corner of the input itself.
+ * @param input The target input element.
+ * @param index The index of the character to calculate for.
+ * @param adjustForScroll Control whether the returned value is adjusted based on scroll position.
+ */
+export const getCharacterCoordinates = (
+  input: HTMLTextAreaElement | HTMLInputElement | null,
+  index: number,
+  adjustForScroll = true
+): Coordinates => {
+  if (!input) return {top: 0, left: 0}
+
+  // word-wrap:break-word breaks the getCaretCoordinates calculations (a bug), and word-wrap has
+  // no effect on input element anyway
+  if (input instanceof HTMLInputElement) input.style.wordWrap = ''
+
+  let coords = getCaretCoordinates(input, index)
+
+  // The library calls parseInt on the computed line-height of the element, failing to account for
+  // the possibility of it being 'normal' (another bug). In that case, fall back to a rough guess
+  // of 1.2 based on MDN: "Desktop browsers use a default value of roughly 1.2".
+  if (isNaN(coords.height)) coords.height = parseInt(getComputedStyle(input).fontSize) * 1.2
+
+  // For some single-line inputs, the rightmost character can be accidentally wrapped even with the
+  // wordWrap fix above. If this happens, go back to the last usable index
+  let adjustedIndex = index
+  while (input instanceof HTMLInputElement && coords.top > coords.height) {
+    coords = getCaretCoordinates(input, --adjustedIndex)
+  }
```

Comment:
Admittedly we shouldn't just be hacking around these bugs in [the library](https://github.com/koddsson/textarea-caret-position). I really should open a PR there and fix them but I've been lazy. The library hasn't been updated in a while but the owner is staff :octocat: so we should be able to get it done. I will definitely try and get to that before we merge this.

Reply:
@koddsson would you mind enabling issues on that repo so I can catalog the bugs there?